docs/DesignDocs/TEECallsRefactoring.md

```diff
@@ -300,218 +282,397 @@ else
+  This ensures that the same ECALL on each per-enclave ECALL table has the same id.
+  For example, the per-enclave ECALL table is defined as follows.
+  ```
+  typedef struct _oe_ecall_id_t
+  {
+     uint64_t id;
+     uint64_t hash;
+  } oe_ecall_id_t;
+  
+  typedef struct _oe_enclave
+  {
+     ...
+     /* Per-edl ecall table. */
+     oe_ecall_id_t ecall_id_table[OE_ECALL_MAX];
+     uint32_t ecall_id_table_size;
+     ...
+  } oe_enclave_t;
+  ```
+  The example of a registration function is as follows.
+  ```
+  oe_result_t oe_host_register_enclave_functions(
+    oe_enclave_t* enclave,
+    const uint64_t* ecall_hashes,
+    uint32_t num_ecalls)
+  {
+      uint32_t i;
+ 
+      oe_update_global_ecall_table(ecall_hashes, num_ecalls);
       
-      /* Check if the index is cached. */
-      if (_caching_table_id == OE_HOST_ECALL_ID_NULL)
-      {
-        _caching_table_id = oe_host_get_ecall_id_by_hash(
-            enclave,
-            _function_hash);
-      }
-
       ...
-      
-      /* Obtain the ecall id from the caching table. */
-      _ecall_id = oe_get_ecall_function_id(enclave, _caching_table_id);
-      if (_ecall_id == -1)
+      for (i = 0; i < _global_ecall_table_size; i++)
       {
-          oe_call_enclave_get_function_id_by_hash(
-              enclave,
-              _function_hash,
-              &_ecall_id);
-          oe_set_ecall_function_id(enclave, _caching_table_id, _ecall_id);
+          enclave->ecall_id_table[i].hash = _global_ecall_table[i];
+          enclave->ecall_id_table[i].id = OE_ECALL_ID_NULL;
       }
-     
-      /* Call enclave function. */
-      if ((_result = oe_call_enclave_function(
-             enclave,
-            (uint64_t) _ecall_id,
-            (uint64_t) _function_hash,
-            _input_buffer,
-            _input_buffer_size,
-            _output_buffer,
-            _output_buffer_size,
-            &_output_bytes_written)) != OE_OK)
-          goto done;
-          
       ...
   }
   ```
-  
+- In the first invocation of an ECALL, the host query the ECALL id by hash from the target enclave
+  (via a special ECALL). Then the host caches the id in the per-enclave ECALL table.
+  For the subsequent invocations, the host uses the cached id from the per-enclave ECALL table (looked
+  up using the same id across different enclaves).
+  These logic can be implemented as part of `oe_call_enclave_function()`, see the following example.
+  ```
+  oe_result_t oe_call_enclave_function(
+      oe_enclave_t* enclave,
+      uint64_t* global_function_id,
+      uint64_t function_hash,
+      const void* input_buffer,
+      size_t input_buffer_size,
+      void* output_buffer,
+      size_t output_buffer_size,
+      size_t* output_bytes_written)
+  {
+    ...
+    uint64_t function_id = OE_ECALL_ID_NULL;
+    static oe_mutex _ecall_lock = OE_H_MUTEX_INITIALIZER;
+
+    ...
+    oe_mutex_lock(&_ecall_lock);
```

Comment:
Consider if all these can be abstracted out into an infrastructure function `oe_get_ecall_function_id`
Something like...
```c
    static uint32_t global_function_id = (uint32_t) -1;
    uint32_t function_id = 0;

    if (oe_get_ecall_function_id(enclave, hash, &global_function_id, &function_id) != OE_OK)
         goto done;
```